hbase-rsgroup/src/test/java/org/apache/hadoop/hbase/rsgroup/TestRSGroupsAdmin2.java

```diff
@@ -459,4 +463,197 @@ public boolean evaluate() throws Exception {
+    for (ServerName server : assignMap.keySet()) {
+      rregion = assignMap.get(server).size() >= 1 && !newGroup.containsServer(server.getAddress()) ?
+          assignMap.get(server).get(0) :
+          null;
+      if (rregion != null) {
+        srcServer = server;
+        break;
+      }
+    }
+    assert srcServer != null;
+    RegionInfo ri = TEST_UTIL.getMiniHBaseCluster().getMaster().getAssignmentManager().
+        getRegionInfo(Bytes.toBytesBinary(rregion));
+    RegionStateNode rsn =
+        TEST_UTIL.getMiniHBaseCluster().getMaster().getAssignmentManager().getRegionStates()
+            .getRegionStateNode(ri);
+    rsn.setState(RegionState.State.SPLITTING);
+
+    // move table to group
+    Thread t2 = new Thread(() -> {
+      LOG.info("thread2 start running, to move regions");
+      try {
+        rsGroupAdmin.moveTables(Sets.newHashSet(tableName), newGroup.getName());
+      } catch (IOException e) {
+        LOG.error("move server error", e);
+      }
+    });
+    t2.start();
+
+    // start thread to recover region state
+    final ServerName ss = srcServer;
+    final String sregion = rregion;
+    AtomicBoolean changed = new AtomicBoolean(false);
+    Thread t1 = new Thread(() -> {
+      LOG.info("thread1 start running, will recover region state");
+      long current = System.currentTimeMillis();
+      while (System.currentTimeMillis() - current <= 50000) {
+        List<RegionInfo> regions = master.getAssignmentManager().getRegionsOnServer(ss);
+        List<RegionInfo> tableRegions = new ArrayList<>();
+        for (RegionInfo regionInfo : regions) {
+          if (regionInfo.getTable().equals(tableName)) {
+            tableRegions.add(regionInfo);
+          }
+        }
+        LOG.debug("server table region size is:{}", tableRegions.size());
+        assert tableRegions.size() >= 1;
+        // when there is exactly one region left, we can determine the move operation encountered
+        // exception caused by the strange region state.
+        if (tableRegions.size() == 1) {
+          assertEquals(tableRegions.get(0).getRegionNameAsString(), sregion);
+          rsn.setState(RegionState.State.OPEN);
+          LOG.info("set region {} state OPEN", sregion);
+          changed.set(true);
+          break;
+        }
+        sleep(5000);
+      }
+    });
+    t1.start();
+
+    t1.join();
+    t2.join();
+
+    TEST_UTIL.waitFor(WAIT_TIMEOUT, new Waiter.Predicate<Exception>() {
+      @Override
+      public boolean evaluate() {
+        if (changed.get()) {
+          boolean serverHasTableRegions = false;
+          for (RegionInfo regionInfo : master.getAssignmentManager().getRegionsOnServer(ss)) {
+            if (regionInfo.getTable().equals(tableName)) {
+              serverHasTableRegions = true;
+              break;
+            }
+          }
+          return !serverHasTableRegions && !rsn.getRegionLocation().equals(ss);
+        }
+        return false;
+      }
+    });
+  }
+
 }
```

Comment:
How about test condition where number of retries is reached and region still can't be moved for some reason?

Reply:
Yes, good idea.

Reply:
I have added testFailedMoveWhenMoveTable testFailedMoveWhenMoveServer to test the case when number of retries is up to limit but region still can't be moved.

Reply:
Overall tests logic is fine, but these still miss a _fail_ call if no exception is thrown by the tested method, otherwise test will pass even if no exception is thrown.